### PR TITLE
COMMERCE-4957 Refactored AddToCart (Button + Wrapper with QuantitySelector)

### PR DESCRIPTION
**Pull request by gianmarcobrunialti**
This is the final PR for the AddToCart component which includes:
- the `AddToCart.js` entry component to decide whether it should display an `AddToCartButton` component along with its quantity-editing `QuantitySelector` component
- the `AddToCartButton.js` component which will replace the current one based on Metal.js/Soy (to be replaced once I'll deliver the ProductCard component)
- the `WrapperWithQuantity.js` component which, by its own name, wraps the `AddToCartButton` component with the `QuantitySelector` component and is in charge for holding the current product's quantity to be selected via the latter and passed on to the former.

By default the `AddToCartButton` activated via `AddToCart` lives as a standalone component and always increases the `quantity` by 1 unit, having no dependency on the `QuantitySelector` (which is not rendered). If appropriate configuration props are passed in `(withQuantity: { ...QuantitySelectorProps }}`, the `WrapperWithQuantity` will be enabled and the quantity will be handled as described above.

Last but not least: this PR also includes a different approach for test cases which make use of precise timers/timeouts within the code and may unpredictably cause failures in CI because of the handling of runtime queues.

### CODE REVIEW

**Comment by gianmarcobrunialti** [comment]
ci:test:sf

**Comment by gianmarcobrunialti** [comment]
ci:test:relevant

File: modules/apps/commerce/commerce-frontend-js/test/utilities/fake_data/add_to_cart_data.js [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+ * details.
+ */
+
+export const accountId = 43936;
+export const channelId = 41005;
+export const currencyCode = 'USD';
+export const orderId = 43939;
+export const productId = 43939;
+export const skuId = 43712;
+export const spritemap = './assets/icons.svg';
```
Comment (gianmarcobrunialti) [inline modules/apps/commerce/commerce-frontend-js/test/utilities/fake_data/add_to_cart_data.js:21]: Note to self: to be removed.

**Comment by gianmarcobrunialti** [comment]
@wincent Would you please like to give it a go with your review? Thanks! :)

File: modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/components/add_to_cart/WrapperWithQuantity.js [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+import classnames from 'classnames';
+import PropTypes from 'prop-types';
+import React, {useState} from 'react';
+
+import QuantitySelector from '../quantity_selector/QuantitySelector';
+
+function WrapperWithQuantity({AddToCartButton, ...props}) {
+	const [currentQuantity, setCurrentQuantity] = useState();
+
+	const onQuantityUpdate = (quantity) => setCurrentQuantity(quantity);
```
Comment (bryceosterhaus) [inline modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/components/add_to_cart/WrapperWithQuantity.js:24]: I don't think you need to re-declare a function here
```suggestion
	const onQuantityUpdate = setCurrentQuantity
```

You can probably just pass `setCurrentQuantity` directly at `onUpdate={setCurrentQuantity}`. So this intermediate variable isn't even necessary.

File: modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/components/quantity_selector/QuantitySelector.js [hunk 3]
```diff
@@ -68,30 +70,32 @@ function QuantitySelector({
...
+		className: classnames({
+			'form-control-lg': large,
+			'quantity-selector': true,
+		}),
+		'data-component-id': componentId,
+		disabled,
+		name,
+		onChange,
+		value: selectedQuantity,
+	};
```
Comment (bryceosterhaus) [inline modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/components/quantity_selector/QuantitySelector.js:83]: This is a really interesting way to handle this, I haven't thought of doing it like this before 👍 👍

File: modules/apps/commerce/commerce-frontend-js/test/components/AddToCart/AddToCart.js [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+
+			const WrapperWithQuantityContainer = Component.container.querySelector(
+				'.add-to-cart'
+			);
+			const QuantitySelectorInputElement = Component.container.querySelector(
+				'input'
+			);
+			const AddToCartButtonElement = Component.container.querySelector(
+				'button'
+			);
```
Comment (bryceosterhaus) [inline modules/apps/commerce/commerce-frontend-js/test/components/AddToCart/AddToCart.js:69]: I don't think you necessarily need to be resetting and nulling the `Component` variable for each test. This test could look a little more simple

```suggestion
			const {container} = render(<AddToCart {...props} />);

			const WrapperWithQuantityContainer =container.querySelector('.add-to-cart');
			const QuantitySelectorInputElement = container.querySelector('input');
			const AddToCartButtonElement = container.querySelector('button');
```

File: modules/apps/commerce/commerce-frontend-js/test/components/AddToCart/AddToCart.js [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+		beforeEach(() => {
+			jest.resetAllMocks();
+
+			window.Liferay.Language.get = jest.fn();
+		});
+
+		afterEach(() => {
+			cleanup();
+
+			Component = null;
```
Comment (bryceosterhaus) [inline modules/apps/commerce/commerce-frontend-js/test/components/AddToCart/AddToCart.js:44]: You don't need to keep this variable from test to test

File: modules/apps/commerce/commerce-frontend-js/test/components/AddToCart/WrapperWithQuantity.js [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+
+		beforeEach(() => {
+			jest.resetAllMocks();
+			jest.useFakeTimers();
+		});
+
+		afterEach(() => {
+			cleanup();
+
+			Component = null;
```
Comment (bryceosterhaus) [inline modules/apps/commerce/commerce-frontend-js/test/components/AddToCart/WrapperWithQuantity.js:53]: Same as my comments from the other test. No need to keep this value around for each test.

File: modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/components/add_to_cart/AddToCart.js [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+
+import AddToCartButton from './AddToCartButton';
+import WrapperWithQuantity from './WrapperWithQuantity';
+
+function AddToCart(props) {
+	return !props?.settings?.withQuantity ? (
+		<AddToCartButton {...props} />
+	) : (
+		<WrapperWithQuantity AddToCartButton={AddToCartButton} {...props} />
+	);
```
Comment (bryceosterhaus) [inline modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/components/add_to_cart/AddToCart.js:26]: This is a little confusing to read, not sure if it would make more sense to check for inverse
```suggestion
	return props?.settings?.withQuantity ? (
	        <WrapperWithQuantity AddToCartButton={AddToCartButton} {...props} />
	) : (
		<AddToCartButton {...props} />
	);
```

**Review by bryceosterhaus — APPROVED** [review]
Looks good to me!

Nothing to stop this from getting pushed through but I left a few comments just on some small nitpicks that I noticed, but those can either be ignored or addressed in a follow up pr.

**Comment by bryceosterhaus** [comment]
Not sure about the test failures. @gianmarcobrunialti, you'll probably want to check them out before forwarding.

File: modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/components/add_to_cart/WrapperWithQuantity.js (hunk 2 above)
Comment (gianmarcobrunialti) [reply modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/components/add_to_cart/WrapperWithQuantity.js:24]: Right, it's definitely neater and I thought of this too. I simply wanted to have an easily debuggable spot within this component instead of leaving everything implicit across the rendering/state changing flow. I'll fix it!

File: modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/components/add_to_cart/AddToCart.js (hunk 7 above)
Comment (gianmarcobrunialti) [reply modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/components/add_to_cart/AddToCart.js:26]: Agreed!

TL;DR: In my mind I was more eager to check against the fewer and steadier _falsy_ possibilities to instantly decide we don't need a WrapperWithQuantity rather than the _truthy_ ones (`withQuantity` may exist, yet still not contain the appropriate props to start up render the component correctly). Psychological security I guess, nothing else, optional checks do the job just fine, as well as PropTypes in the QuantitySelector.

**Comment by gianmarcobrunialti** [comment]
ci:test:sf

**Comment by gianmarcobrunialti** [comment]
ci:test:relevant

**Comment by gianmarcobrunialti** [comment]
Thanks a lot for your review! @bryceosterhaus 💯

File: modules/apps/commerce/commerce-frontend-js/test/dev/components/AddToCart.js [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+			options: '[]',
+			skuId: 43712,
+		},
+		orderId: 43939,
+		quantity: 1,
+		settings: {
+			block: false,
+			disabled: false,
+			iconOnly: false,
+			withQuantity: /* false, */ {
```
Comment (wincent) [inline modules/apps/commerce/commerce-frontend-js/test/dev/components/AddToCart.js:37]: What's this comment for?

**Comment by marco-leo** [comment]
ci:test:relevant

**Comment by marco-leo** [comment]
Hey @gianmarcobrunialti the failures looks unique to this pull

**Comment by gianmarcobrunialti** [comment]
ci:test:relevant

**Comment by gianmarcobrunialti** [comment]
ci:test:sf

File: modules/apps/commerce/commerce-frontend-js/test/dev/components/AddToCart.js (hunk 8 above)
Comment (gianmarcobrunialti) [reply modules/apps/commerce/commerce-frontend-js/test/dev/components/AddToCart.js:37]: Hey @wincent , alternative dev-env test setting, but I know the policy about comments, I simply forgot to remove it. As I believe the js-unit tests are finally passing and the SF should run smoothly, may I clean this up in a subsequent PR (it's gonna come right on top of this)? Promise 🙏

Comment (wincent) [reply modules/apps/commerce/commerce-frontend-js/test/dev/components/AddToCart.js:37]: Sure. Just be aware that if Brian sees it he might reject it. 🤞

Comment (gianmarcobrunialti) [reply modules/apps/commerce/commerce-frontend-js/test/dev/components/AddToCart.js:37]: 🤞

**Comment by ivicac** [comment]
ci:forward

**Comment by ivicac** [comment]
ci:forward

**Review by ivicac — APPROVED** [review]

**Comment by ivicac** [comment]
ci:stop

**Comment by ivicac** [comment]
ci:forward
